### ISSUE TITLE
Is -handleDeepLink: handling links that don't belong to it?

### ISSUE THREAD

**Issue by hsoi** [BranchMetrics/ios-branch-deep-linking#370]
# Questions

Is `-[Branch handleDeepLink:]` handling links that don't belong to it?

Is `-handleDeepLink:` supposed to handle links that don't belong to it?

Are we supposed to determine if the link is a Branch link or not before passing it along to `-handleDeepLink:`?

# Situation

In my app's `-application:openURL:sourceApplication:annotation:` implementation I placed handlers for both Branch and Facebook (we use Facebook Login support). [According to the Branch documentation](https://dev.branch.io/getting-started/sdk-integration-guide/guide/ios/#handle-incoming-links):

```
// Respond to URI scheme links
- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {
    // pass the url to the handle deep link call
    [[Branch getInstance] handleDeepLink:url];

    // do other deep link routing for the Facebook SDK, Pinterest SDK, etc
    return YES;
}
```

My difference is I check if `-[Branch handleDeepLink:]` returns if it was handled or not and if not, then pass it along to Facebook's SDK. The way the documentation reads, the way the API is designed, the impression I am given is that `-[Branch handleDeepLink:]` will take the URL given and if it's a Branch URL will handle it (and return YES), else if it's not a Branch link will not handle it (and return NO).

But this doesn't appear to happen.

Here's the implementation as of this writing (v0.12.0 is what we used, but v0.12.2 -- current release -- seems to have it as well):

```
- (BOOL)handleDeepLink:(NSURL *)url {
    BOOL handled = NO;
    if (url) {
        //always save the incoming url in the preferenceHelper in the externalIntentURI field
        self.preferenceHelper.externalIntentURI = [url absoluteString];

        NSString *query = [url fragment];
        if (!query) {
            query = [url query];
        }
        
        NSDictionary *params = [BNCEncodingUtils decodeQueryStringToDictionary:query];
        if (params[@"link_click_id"]) {
            handled = YES;
            self.preferenceHelper.linkClickIdentifier = params[@"link_click_id"];
        }
    }
    
    [self initUserSessionAndCallCallback:YES];
    
    return handled;
}
```

It seems there is work done to determine if `url` is a Branch URL or not by looking for the `link_click_id`. 

Thing is, `-initUserSessionAndCallCallback:` seems to be called regardless -- if `url` is `nil`, it's called. If the URL is not a Branch URL (`handled == NO`), it's called.

That does not seem right to me. It seems to me that we should only be initing the user session and calling the callback if `handled == YES`. Or perhaps you always init the user session, but you should only call the callback if `handled == YES` -- I'm not intimately familiar with the SDK's inner workings to know which would be proper. But either way, when the URL is not a Branch URL, the callback is being called, and I don't believe that's proper.

In my case, my app get the Facebook auth URL, and then my Branch handler is being invoked! When one is within the handler, there's no way to determine what the original/invoking URL is -- and in my case, I was passed some sort of older, stale/prior URL query data -- so it SEEMED to be a branch URL that I should be handling, but well... I'm responding to a Facebook auth so there's ZERO reason that any sort of Branch handling should be going on.

# Solution?

Given what I see from Branch in terms of the documentation (both on the website and in the header file) and in terms of how the API is written and implemented, I get the impression that `-handleDeepLink:` is supposed to only invoke the callback if it's an actual Branch URL (i.e. `handled == YES`). That it's supposed to be callable for any URL (even nil), and it will "do the right thing". The only thing the caller should be responsible for is ensuring they check the return value and proceed accordingly.

If however it's the callers responsibility to only pass Branch URLs along -- which seems less optimal and like a maintenance problem -- then documentation needs to be updated that it is the caller's responsibility to determine this, and hopefully with guidance as to how this is to be accurately done.

Or if it's unsafe to pass "any old URL" to this API, it needs to be documented as such.

Thank you.

**Comment by aaustin** [comment]
Hey @hsoi - Thanks so much for your thorough note.

Before responding to your questions, let me address the concern about the call to `[self initUserSessionAndCallCallback:YES];` regardless of whether the `link_click_id` is there. We initialize a Branch session every single time the app is opened, regardless of whether a Branch link was clicked or not. We do this because it's quite possible that we must resort to our matching mechanism of deep linking, where we use our fingerprints to determine if a link was clicked. You can read more about that [here](https://dev.branch.io/getting-started/matching-accuracy/guide/). We use the `link_click_id` if it's present, since it can be more accurate than our fingerprinting _in certain circumstances_. However, please expect a Branch session to initialize every time the app is opened, meaning the deep link routing block will be called 100% of the time, optionally containing Branch link data.

And now to your questions:

> Is -[Branch handleDeepLink:] handling links that don't belong to it?

We will receive URI schemes that don't belong to Branch in this method but not act on them if `link_click_id` is not present. Soon, however, we will start reporting stats on non-Branch referrers on the Branch dashboard. So we're collecting this info on your behalf for future analytics. Coming soon.

> Is -handleDeepLink: supposed to handle links that don't belong to it?

Yes. It is.

> Are we supposed to determine if the link is a Branch link or not before passing it along to -handleDeepLink:?

No. Don't worry about it. We don't muck around with other links if they are passed to Branch and will only return `YES` if we know the link is ours. However!! There are other deep link receives (like Facebook's), that actually modify and change the URL passed it it. We've found that Facebook (maybe maliciously) strips out the `link_click_id` from our links. The moral is: please call Branch before you call your other handlers.

Hope this helps!

**Comment by hsoi** [comment]
Hi @aaustin and thank you for the reply.

First, it makes sense that you'd initialize the session every time. But after that, things are confusing.

From what I gather you're saying, the deep link handler is called EVERY time, regardless -- and that is by design. OK, fair enough.  But you also say it optionally contains Branch link data.

Under what circumstances does it -- and doesn't it?

Should the handler contain Branch link data if the link is not a Branch link?

So my gathering then is if say the URL is not a Branch URL, the handler should not receive Branch link data.  I mean, when my handler is invoked, I need to obviously be able to sniff that param data to know "OK, handle this link this way; handle that link that way". So I would think if the URL was NOT a Branch link, there should be no branch link data in there at all... maybe `param` is empty or `nil` or some such, yes?

Because in my case, well... my app gets launched because the user tapped on a Branch link, so my app launches, there's Branch link data stored somewhere. Then while my app is handling it the user has to authenticate, they use Facebook Login to do so, when the FB auth URL comes through I hit the handler per my above description... and for whatever reason when I get into my Branch handler the params data is full of the original Branch URL data that launched the app. And so my app tries to handle the Branch URL all over again (while it's already in process of doing so).

That doesn't seem right.

Because of this, it got our app into a very wonky state, and so while originally we first handled Branch then passed to Facebook if Branch didn't handle, we had to work around it for now by first passing to Facebook then passing to Branch if not handled.  I'd like to think there shouldn't be a link-stripping problem here because we pass our own immutable instance of the `NSURL` each time... tho maybe Facebook is doing something untoward under the hood. Still, we had to do this to avoid the crash.


So, while I can see what y'all are trying to do, something still doesn't seem right here. Hoping you can clarify.  Thank you.

**Comment by hsoi** [comment]
To say it another way...

I would think that each time the deep link handler was invoked, the params would be for THAT invocation of the handler.  Well, it seems that's NOT the case -- at least in terms of what I encountered. That for the invocation of the handler for that Facebook URL, I was receiving the params for the prior Branch URL (the one that actually launched the app and started the whole process I was doing). 

That doesn't seem right.

But there may be something more at play here.

**Comment by aaustin** [comment]
Ah - that definitely appears to be unexpected behavior. Let me put it in a timeline so we're both on the same page:
0. Click a Branch link
1. App opens and deep link async block returns with deep link data
2. Click login with Facebook
3. Facebook app opens
4. Facebook deep links back to your app
5. Deep link async block returns original Branch deep link data

This definitely shouldn't happen because when the Facebook app opens, your app moves to the background and our session management code will automatically close the session and clear the deep link data out. If you're using `[branch setDebug];` before you call initSession, it will log all requests. Can you please confirm that you see a call to /v1/close when the Facebook app opens?

Assuming that it does, I bet I know what the issue is. Unfortunately, the deferred deep linking technique that I refer to above as 'fingerprinting' has some downfalls that you must be wary of for testing. We do everything we can to mitigate them but it can cause some confusion like you're experiencing if you're not careful.

Here's what probably happened:
1. You pasted a link into Safari which redirected to the App Store because it doesn't support Universal Links.
   - this generated an outstanding click/pending match
2. You clicked a link from Slack or Messages which correctly triggered a Universal Link
3. The app opened and you received the deep link params from the click in 2
   - note that there is still an outstanding click/pending match from 1
4. Facebook closed the Branch session and reopened it by deep linking back to the ap
5. Branch initializes a new session and it matches to the outstanding click in 1
6. Deep link parameters returned and confusion ensues.

One quick way to mitigate this and verify it's the root cause it to open and close the app before triggering the Facebook login to clear out any previous outstanding matches from testing. Then trigger the login and no data should come through.

Let me know your thoughts

**Comment by hsoi** [comment]
Hi @aaustin 

Timeline is close. Here's what I did just now to reproduce it again.

- obtained the v1.0.0 code for our app (that which has the problem)
- deleted the app off my phone
- added the `setDebug()` call before initialization so I could watch more of what's going on.
- built the app, ran it (via Xcode debugger)
- the app ran and launched, sat idle, all seems good and clear.
- left the app, entered Messages.app, tapped on a URL that deep links back into our app
- the URL resolved and it switched back to my app. I see the Branch handler invoked and the params contain the data I'd expect for that URL.
- the user proceeds in our app's flow and now need to log in in order to work with the stuff linked to. So we present our login UI, and the user taps on the "Login with Facebook" button.
- NOTE! We are using the FB SDK here for the login support, but it does NOT actually leave our app and launch the FB app (we're using FB SDK Core Kit and FB SDK Login Kit 4.11.0). What FB does is present a `SFSafariViewController` and the auth happens in there -- my app is NEVER left, never backgrounded. 
- Do the login... the FB auth URL comes into the app
- Because we give Branch the chance to handle the URL first, our Branch handler is invoked - and the params to the handler are the same as the params from the prior URL call (as referenced above).

and then everything starts to go wonky

I eventually see a call to `/v1/close` but not until WELL afterwards.  And again, since we never leave our app, there's no backgrounding, thus no auto-session closing on your end.

----

So I tried something else per a suggestion you gave.

I did most of the above same process. I clicked on the URL, got into my app and then at the landing page in my app where the user has to choose to log in.  So the Branch URL has been brought into the app and essentially handled (but no api close yet).  I then explicitly hit the Home button to switch my app to the background (before I log in). When I did this, I saw the call to the `api/close`.

I then bring my app to the foreground. The Branch handler was invoked, but just had some general stuff in there. It was NOT anything from before, and because it lacked anything my app was explicitly looking for (i.e. our content identifiers), the handler exited without a problem.

And so I proceed to Login via Facebook. Again SFSafariViewController. I auth, the auth URL comes in... and while the Branch handler was invoked, all it gave as params was:

```
[DeepLinkRouter.swift:25] branchDeepLinkHandler > params: ["+clicked_branch_link": 0, "+is_first_session": 0]
```

So 👏 

And things proceeded along just fine.

-----

So yeah, I guess the problem is because we don't actually leave the app thus the session never gets closed, so the next invocation of the handler has the prior URL's data... which isn't right.

Does this help?

Where to go from here?

**Comment by aaustin** [comment]
Ahhh. The openUrl call is coming from inside the house! 

This is definitely new behavior it seems, and our SDK isn't properly set to receive these phony Facebook openUrl calls. We'll need to make a bit of a patch here to accept these. Give me a couple days to sort this out.

In the mean time, you could probably just add a hack like so:
```if (![url hasPrefix:@"fb"]) { [[Branch getInstance] handleDeepLink:url]; } ```

**Comment by hsoi** [comment]
Ah! It gets even better. 😆 

Well, because this made everything go wonky for us, the "patch" we did was to switch the ordering around, calling the FB handler first, then if it didn't handle calling the Branch handler. So we've already got a patch in the wild. Given all that you've said tho, a patch like yours seems likely better.

That said, since I can't rush out yet another patch... I'll wait to see how timing goes on things relative to the next update we release. If you have an updated SDK, we'll just update to that and see how it goes (and switch the handler invocation order back), else switch around to the above workaround.

I guess just keep this Issue open and updated as things go along.

Thanx for hanging with me on this! 👍
